Probabprime

yaalen opened the pull request:
Miller-Rabin primality test for BPSW algorithm.

wbhart commented:
Hi Elena,

sorry for the long delay. I was travelling a lot recently.

I have taken a closer look at your code. Here are some comments:

1) Can you add documentation for the new functions in ulong_extras/doc/ulong_extras.txt

2) On line 41 of is_probabprime_miller_rabin the comment should presumably have gcd(n, i) not gcd(n, b), Also occurs later on lines 60 and 64.

3) What happens if n < i? Does the powmod still work correctly?

4) It would be cheaper to precompute an inverse of n. Then you could use it in powmod_precomp/preinv. Also the second powmod in each case could become a mulmod_precomp/preinv, which might be cheaper.

5) The second loop in each case should return 0 if rem == 1 should it not? You want this to happen as soon as possible rather than wait until all iterations have completed.

6) On line 47 of n_is_probabprime_BPSW_miller_rabin I don't understand the change to if ((n % 10) != 9). That doesn't seem to be equivalent to the test we had there before. Can you explain this change?

7) In n_is_probabprime_BPSW_miller_rabin you only call miller_rabin with base 2. This is sufficient, but is it optimal? It would be nice to know if using additional bases might be faster. E.g. maybe using base 2 and then base 3 is faster. Probably not, but it is worth testing both for random primes that are 3 or 7 mod 10 and for random composites that are 3 or 7 mod 10. I would expect the primes to slow down and the composites to speed up. But the question is, what is the tradeoff?

3) The gcd test on line 43 only needs to be run if i > 2, since n is guaranteed odd by this point.

4) There are some spaces missing on line 50.

5) On line 52, may as well use postincrement in line with the rest of flint.

6) Why have two variables for both p and for q when only one is ever used.

yaalen commented:
1) *Can you add documentation for the new functions in ulong_extras/doc/ulong_extras.txt*
done!
Also added some spaces in ../../fmpz_mat/doc/fmpz_mat.txt due to the
Parse exception:
Encountered malformed input near line 333
in file ../../fmpz_mat/doc/fmpz_mat.txt

2) *On line 41 of is_probabprime_miller_rabin the comment should presumably have gcd(n, i)
not gcd(n, b), Also occurs later on lines 60 and 64. Or you could just change the variable
name i to b throughout.*
Done!

3) *What happens if n < i? Does the powmod still work correctly?*
I'm not sure I've understand your question exactly. Do you mean the outstanding powmod by itself or how we can use it inside n_is_probabprime_miller_rabin function?
In the MR test as it is we randomly select a number i such that 0 < i < n to test n is probabprime.
So there is a restriction on i and n and we probably can assume i < n. Is that right?

4) *It would be cheaper to precompute an inverse of n. Then you could use it in powmod_precomp/preinv.
Also the second powmod in each case could become a mulmod_precomp/preinv, which might be cheaper.*
Surely, thank you for this note. My idea was to use inlined n_powmod(n_powmod2) functions instead of _precomp(_preinv) versions. Now I've compared timings for both versions and saw that this idea was wrong.

5) *The second loop in each case should return 0 if rem == 1 should it not? You want this to happen
 as soon as possible rather than wait until all iterations have completed.*
Of course, you are right, it should and it will be faster! fixed.

6) *On line 47 of n_is_probabprime_BPSW_miller_rabin I don't understand the change to if
((n % 10) != 9). That doesn't seem to be equivalent to the test we had there before.
Can you explain this change?*
I had played a little with this tests(just because I wondered how much the timing's difference can be) and then I've forgot to rollback this test to upstream version before commit, sorry about that
The upstream test was better than 'new' one. So it should be restored to the origin version.
(I've done this already.)

7) *In n_is_probabprime_BPSW_miller_rabin you only call miller_rabin with base 2.
This is sufficient, but is it optimal? It would be nice to know if using additional bases
might be faster. E.g. maybe using base 2 and then base 3 is faster. Probably not, but it is worth
testing both for random primes that are 3 or 7 mod 10 and for random composites that are 3 or 7 mod 10. I would expect the primes to slow down and the composites to speed up. But the question is, what is the tradeoff?*
Thank you for this advice, I'll try to collect some statistics about it.

8) *The gcd test on line 43 only needs to be run if i > 2, since n is guaranteed odd by this point.*
9) *There are some spaces missing on line 50.*
10) *On line 52, may as well use postincrement in line with the rest of flint.*
fixed.

11) *Why have two variables for both p and for q when only one is ever used.*
q is used later in the code, first use is on line 56. Certainly, there is some redundancy on using both tmp and q variables. I think this piece of code seems to be more readable this way. I can replace any of q or tmp with other one, of course. Please, tell me if you'll find such replacement will be relevant.

12) *To check that everything is working ok, I highly recommend comparing the output of
this code against our n_is_prime function for n = 0, 1, 2, ...., 2^32-1 and the for a few hours
 with random primes up to 2^64. Make sure you print the values if there is a failure so you can
 track it down.*
In progress. Thanks for the tip!

wbhart commented:
Thanks for these changes.

I would definitely prefer if you could remove the additional variables that
end up becoming p and q. This is just a preference obviously. In a sense,
you are telling the compiler to allocate additional space for variables
when you have more of them (of course it might have optimisations that
figure out it can do with less space).

I realise it's not a huge issue here, but in recursive functions, for
example, it can become an issue. So I always personally make a practice of
keeping the number of variables low, so long as it doesn't make the code
unreadable.

I guess number 3 is not a problem if you have documented clearly that the
function requires 0 < i < n. You are right that any higher level function
will always choose random values of i in that range.

In some other primality test functions in flint we've dealt with the case i
> n by reducing i modulo n if it is too large. But maybe this function is
low enough level that this isn't important. So long as the requirement is
document it should be ok, and so long as the code can't increase i so that
it becomes >= n, otherwise the modular reduction code will no longer work.

This brings me to another point. Should there be a function which takes a
number of iterations and either chooses bases in order, or randomly (though
this is likely to be too expensive) up to some limit, and calls the
miller_rabin code for those bases? Somehow I think that code should be
called n_is_probabprime_miller_rabin, which means that perhaps this
function should be called something else.

You might also like to look at n_is_strong_probabprime_precomp/preinv,
since unless I am mistaken, those are essentially doing what your current
miller_rabin code is doing (I didn't check this carefully). So in fact, the
outer miller_rabin function is probably more important to have at the
moment, since I think this is genuinely missing. It might also make it
easier to experiment with in the BPSW code.

I'll wait until you have a sense of whether 7 and 12 improve performance
and have looked at the above issues before merging.

Thanks again for your contribution!

Bill.

On 8 April 2015 at 01:20, yaalen <notifications@github.com> wrote:

> 1) *Can you add documentation for the new functions in
> ulong_extras/doc/ulong_extras.txt*
> done!
> Also added some spaces in ../../fmpz_mat/doc/fmpz_mat.txt due to the
> Parse exception:
> Encountered malformed input near line 333
> in file ../../fmpz_mat/doc/fmpz_mat.txt
>
> 2)
>
> *On line 41 of is_probabprime_miller_rabin the comment should presumably
> have gcd(n, i) not gcd(n, b), Also occurs later on lines 60 and 64. Or you
> could just change the variable name i to b throughout.*
> Done!
>
> 3) *What happens if n < i? Does the powmod still work correctly?*
> I'm not sure I've understand your question exactly. Do you mean the
> outstanding powmod by itself or how we can use it inside
> n_is_probabprime_miller_rabin function?
> In the MR test as it is we randomly select a number i such that 0 < i < n
> to test n is probabprime.
> So there is a restriction on i and n and we probably can assume i < n. Is
> that right?
>
> 4)
> *It would be cheaper to precompute an inverse of n. Then you could use it
> in powmod_precomp/preinv. Also the second powmod in each case could become
> a mulmod_precomp/preinv, which might be cheaper.*
> Surely, thank you for this note. My idea was to use inlined
> n_powmod(n_powmod2) functions instead of _precomp(_preinv) versions. Now
> I've compared timings for both versions and saw that this idea was wrong.
>
> 5)
> *The second loop in each case should return 0 if rem == 1 should it not?
> You want this to happen as soon as possible rather than wait until all
> iterations have completed.*
> Of course, you are right, it should and it will be faster! fixed.
>
> 6)
>
> *On line 47 of n_is_probabprime_BPSW_miller_rabin I don't understand the
> change to if ((n % 10) != 9). That doesn't seem to be equivalent to the
> test we had there before. Can you explain this change?*
> I had played a little with this tests(just because I wondered how much the
> timing's difference can be) and then I've forgot to rollback this test to
> upstream version before commit, sorry about that
> The upstream test was better than 'new' one. So it should be restored to
> the origin version.
> (I've done this already.)
>
> 7)
>
>
> *In n_is_probabprime_BPSW_miller_rabin you only call miller_rabin with
> base 2. This is sufficient, but is it optimal? It would be nice to know if
> using additional bases might be faster. E.g. maybe using base 2 and then
> base 3 is faster. Probably not, but it is worth testing both for random
> primes that are 3 or 7 mod 10 and for random composites that are 3 or 7 mod
> 10. I would expect the primes to slow down and the composites to speed up.
> But the question is, what is the tradeoff?*
> Thank you for this advice, I'll try to collect some statistics about it.
>
> 8) *The gcd test on line 43 only needs to be run if i > 2, since n is
> guaranteed odd by this point.*
> 9) *There are some spaces missing on line 50.*
> 10) *On line 52, may as well use postincrement in line with the rest of
> flint.*
> fixed.
>
> 11) *Why have two variables for both p and for q when only one is ever
> used.*
> q is used later in the code, first use is on line 56. Certainly, there is
> some redundancy on using both tmp and q variables. I think this piece of
> code seems to be more readable this way. I can replace any of q or tmp with
> other one, of course. Please, tell me if you'll find such replacement will
> be relevant.
>
> 12)
>
>
> *To check that everything is working ok, I highly recommend comparing the
> output of this code against our n_is_prime function for n = 0, 1, 2, ....,
> 2^32-1 and the for a few hours with random primes up to 2^64. Make sure you
> print the values if there is a failure so you can track it down.*
> In progress. Thanks for the tip!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/wbhart/flint2/pull/129#issuecomment-90758264>.
>
